x/precisebank/keeper/send.go

```diff
@@ -19,23 +32,291 @@ func (k Keeper) SendCoins(
+// Depending on the fractional balance and the amount being transferred:
+// Sender account:
+//  1. Arithmetic borrow 1 integer equivalent amount of fractional coins if the
+//     fractional balance is insufficient.
+//  2. No borrow if fractional balance is sufficient.
+//
+// Receiver account:
+//  1. Arithmetic carry 1 integer equivalent amount of fractional coins if the
+//     received amount exceeds max fractional balance.
+//  2. No carry if received amount does not exceed max fractional balance.
+//
+// The 4 cases are:
+// 1. Sender borrow, receiver carry
+// 2. Sender borrow, NO receiver carry
+// 3. NO sender borrow, receiver carry
+// 4. NO sender borrow, NO receiver carry
+//
+// Truth table:
+// | Sender Borrow | Receiver Carry | Direct Transfer |
+// | --------------|----------------|-----------------|
+// | T             | T              | T               |
+// | T             | F              | F               |
+// | F             | T              | F               |
+// | F             | F              | F               |
+func (k Keeper) sendExtendedCoins(
+	ctx sdk.Context,
+	from, to sdk.AccAddress,
+	amt sdkmath.Int,
+) error {
+	// Sufficient balance check is done by bankkeeper.SendCoins(), for both
+	// integer and fractional-only sends. E.g. If fractional balance is
+	// insufficient, it will still incur a integer borrow which will fail if the
+	// sender does not have sufficient integer balance.
+
+	// Load required state: Account old balances
+	senderFracBal := k.GetFractionalBalance(ctx, from)
+	recipientFracBal := k.GetFractionalBalance(ctx, to)
+
+	// -------------------------------------------------------------------------
+	// Pure stateless calculations
+	integerAmt := amt.Quo(types.ConversionFactor())
+	fractionalAmt := amt.Mod(types.ConversionFactor())
+
+	// Account new fractional balances
+	senderNewFracBal, senderNeedsBorrow := subFromFractionalBalance(senderFracBal, fractionalAmt)
+	recipientNewFracBal, recipientNeedsCarry := addToFractionalBalance(recipientFracBal, fractionalAmt)
+
+	// Case #1: Sender borrow, recipient carry
+	if senderNeedsBorrow && recipientNeedsCarry {
+		// Can directly transfer borrow/carry - increase the direct transfer by 1
+		integerAmt = integerAmt.AddRaw(1)
+	}
+
+	// -------------------------------------------------------------------------
+	// Stateful operations for transfers
+
+	// This includes ALL transfers of >= conversionFactor AND Case #1
+	// Full integer amount transfer, including direct transfer of borrow/carry
+	// if any.
+	if integerAmt.IsPositive() {
+		transferCoin := sdk.NewCoin(types.IntegerCoinDenom, integerAmt)
+		if err := k.bk.SendCoins(ctx, from, to, sdk.NewCoins(transferCoin)); err != nil {
+			return k.wrapError(ctx, from, amt, err)
+		}
+	}
+
+	// Case #2: Sender borrow, NO recipient carry
+	// Sender borrows by transferring 1 integer amount to reserve to account for
+	// lack of fractional balance.
+	if senderNeedsBorrow && !recipientNeedsCarry {
```

Comment:
Using the bools directly instead of assigning a new `transferBorrowToReserve` - I think more clear this way